hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileReaderImpl.java

```diff
@@ -0,0 +1,284 @@
+    BlockIndexEntry blockIndexEntry = metaBlockIndexEntryMap.get(new UTF8StringKey(metaBlockName));
+    if (blockIndexEntry == null) {
+      return Option.empty();
+    }
+    HFileBlockReader blockReader = new HFileBlockReader(
+        context, stream, blockIndexEntry.getOffset(),
+        blockIndexEntry.getOffset() + blockIndexEntry.getSize());
+    HFileMetaBlock block = (HFileMetaBlock) blockReader.nextBlock(HFileBlockType.META);
+    return Option.of(block.readContent());
+  }
+
+  @Override
+  public long getNumKeyValueEntries() {
+    try {
+      initializeMetadata();
+      return trailer.getNumKeyValueEntries();
+    } catch (IOException e) {
+      throw new RuntimeException("Cannot read HFile", e);
+    }
+  }
+
+  @Override
+  public int seekTo(Key key) throws IOException {
+    Option<KeyValue> currentKeyValue = getKeyValue();
+    if (!currentKeyValue.isPresent()) {
+      return SEEK_TO_EOF;
+    }
+    int compareCurrent = key.compareTo(currentKeyValue.get().getKey());
+    if (compareCurrent > 0) {
+      if (currentDataBlockEntry.get().getNextBlockKey().isPresent()) {
+        int comparedNextBlockFirstKey =
+            key.compareTo(currentDataBlockEntry.get().getNextBlockKey().get());
+        if (comparedNextBlockFirstKey >= 0) {
+          // Searches the block that may contain the lookup key based the starting keys of
+          // all blocks (sorted in the TreeMap of block index entries), using binary search.
+          // The result contains the greatest key less than or equal to the given key,
+          // or null if there is no such key.
+
+          Map.Entry<Key, BlockIndexEntry> floorEntry = dataBlockIndexEntryMap.floorEntry(key);
+          if (floorEntry == null) {
+            // Key smaller than the start key of the first block
+            return SEEK_TO_BACKWARDS;
+          }
+          currentDataBlockEntry = Option.of(floorEntry.getValue());
+          currentDataBlock = Option.empty();
+          currentPos.setOffset(
+              (int) currentDataBlockEntry.get().getOffset() + HFILEBLOCK_HEADER_SIZE);
+        }
+      }
+      if (!currentDataBlockEntry.get().getNextBlockKey().isPresent()) {
+        // This is the last data block.  Check against the last key.
+        if (fileInfo.getLastKey().isPresent()) {
+          int comparedLastKey = key.compareTo(fileInfo.getLastKey().get());
+          if (comparedLastKey > 0) {
+            currentDataBlockEntry = Option.empty();
+            currentDataBlock = Option.empty();
+            currentPos.setEof();
+            return SEEK_TO_EOF;
+          }
+        }
+      }
+
+      if (!currentDataBlock.isPresent()) {
+        currentDataBlock = Option.of(instantiateHFileDataBlock(currentDataBlockEntry.get()));
+      }
+
+      return currentDataBlock.get()
+          .seekTo(currentPos, key, (int) currentDataBlockEntry.get().getOffset());
+    }
+    if (compareCurrent == 0) {
+      return SEEK_TO_FOUND;
+    }
+    // Backward seek not supported
```

Comment:
not supported? is that just a comment or should we throw an exception if unsupported

Reply:
I've added the logic to detect backward seeks, separate from the keys less than the first key.  In that case, exception is thrown.